service/mapper/forest.go

```diff
@@ -0,0 +1,66 @@
+// Copyright 2021 Optakt Labs OÜ
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not
+// use this file except in compliance with the License. You may obtain a copy of
+// the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations under
+// the License.
+
+package mapper
+
+import (
+	"github.com/onflow/flow-go/ledger"
+	"github.com/onflow/flow-go/ledger/complete/mtrie/trie"
+	"github.com/onflow/flow-go/model/flow"
+)
+
+type Step struct {
+	prev  flow.StateCommitment
+	tree  *trie.MTrie
+	paths []ledger.Path
+}
+
+type Forest struct {
+	trees map[flow.StateCommitment]Step
+}
+
+func NewForest() *Forest {
+	f := Forest{
+		trees: make(map[flow.StateCommitment]Step),
+	}
+	return &f
+}
+
+func (f *Forest) Load(commit flow.StateCommitment) (flow.StateCommitment, *trie.MTrie, []ledger.Path, bool) {
+	step, ok := f.trees[commit]
+	return step.prev, step.tree, step.paths, ok
+}
+
+func (f *Forest) Save(prev flow.StateCommitment, tree *trie.MTrie, paths []ledger.Path) bool {
+	commit := flow.StateCommitment(tree.RootHash())
+	_, ok := f.trees[commit]
+	if ok {
+		return false
+	}
```

Comment:
Why this check/return value? Is there ever a case where we might accidentally overwrite a state commitment's tree? 🤔